### DISCUSSION TITLE
[HELP] Using BDFR by running individual pre-configured script files.

### DISCUSSION THREAD

**Discussion by StuffonGithub** [Serene-Arc/bulk-downloader-for-reddit#304] [Q&A] [answered]
- [x] I am running the latest version of BDfR
- [x] I have read the [Opening an issue](README.md#configuration)
  
## Description
Hi, I need a little help if that's okay, because I'm lost and have no idea what to do next. Basically just want to ask how I can run BDFR through a .ps1 file?

Back in v1, I used to be able to simply call on the .exe file (even if the .ps1 file is saved a couple folders deep) by running variations of `..\bulk-downloader-for-reddit.exe +commands`. With v2 however, I tried a similar approach by trying to call the .py script using `python3 C:\BDFR\filename.py +commands`, but I can't get it to work. Although, I don't know the right file to call so I just tried every .py file in the BDFR directory but with the same results. 😅

Additionally, running the command from anywhere but the main BDFRv2 folder causes the downloader to revert back to the original naming scheme (see #295) ignoring the manual changes I made based on #296. Not sure if this will be fixed if I call the correct file but figured I mention it anyway just in case.

Thank you and sorry for opening yet another issue. 😁

**Comment by aliparlakci** [comment]
First of all, be sure that `python --version` spits out Python 3.9.x. Then, do `python -m pip install bdfr`

Finally change the `bulk-downloader-for-reddit.exe` with `python -m bdfr` in your .ps1 scripts.

For the second matter, you should open a new issue and provide more detail.

**Reply by StuffonGithub** [reply]
Thanks, @aliparlakci. Yes, I can confirm I have Python 3.9.4 and I have installed the new BDFR version by following the instructions in the readme (with the exception of the two files that I changed based on #296).

The reason I tried calling for a python script instead of using the default `python -m bdfr` command in the .ps1 files is because BDFR just closes a second after running it using the default. Although, the same thing is happening when I tried a direct call to a .py script. 😔

**Reply by StuffonGithub** [reply]
I just read your comment in #291:

> Also, this installation is global. That is, you can run your .bat files in any directory without having to have project files next to it.

So, there must be something wrong in my setup if that's the case. I'll try to figure it out as well.

**Comment by Serene-Arc** [comment, marked as answer]
@StuffonGithub The second issue suggests that you have a different version installed in your PATH. If you are in the BDFR folder, then that version will always take precedence over an installed one. I would suggest reinstalling it with `python3 -m pip install .` or installing an editable version that will always update automatically with any changes with `python3 -m pip install -e .`.

**Reply by StuffonGithub** [reply]
@Serene-Arc That was my thought as well but I couldn't find any bdfr folder in any of the directories listed on my PATH that's why I was curious.

I did try to reinstall but all I got was a `Requirement already satisfied:` response on all packages.

**Reply by Serene-Arc** [reply]
It's installed with all the other python packages. You can do `python3 -m pip install --force-reinstall .` to do it again, since you change the code without taking a corresponding commit, our version manager might not recognise that the code has changed since it uses git to version the package

**Reply by StuffonGithub** [reply]
I did `--force-install` and it redid the whole thing. Thanks! I will keep that for future reference. But it's still using the UNIX timestamp when I tested. I wandered around and saw #296 is in `development` and not in `master`. Could this be why?

**Reply by Serene-Arc** [reply]
Oh! Yes that will be why, development is where all the recent commits should be. master is pretty much where the last release was and no further

**Reply by StuffonGithub** [reply]
I see. Is there a way to install the development version or I can only wait for the release? It's not too important right now since I can run it on the root no problem and I can't get my .ps1 files to run anyway but just for my own edification. 😁

**Reply by Serene-Arc** [reply]
Yes, if you are using git. If you are, go to step 2.

1. Run `git clone https://github.com/aliparlakci/bulk-downloader-for-reddit.git` inside a folder
2. Go into the bulk-downloader-for-reddit folder
3. Run the command `git checkout development` to get the code currently on the development branch
  A. If you are not cloning the repository now for the first time, run `git pull` to get the latest code
4. Run `python3 -m pip install -e .` to install an editable package i.e. one that will reflect any changes in the code in that directory immediately but can still be run from anywhere in the system (assuming that Python is in your PATH, of course)

**Reply by StuffonGithub** [reply]
@Serene-Arc
Thanks a lot for dumbing down Git for me. I always thought it required a lot more extra stuff I needed to remember compared to just downloading the zip archive from a project so I never thought to try it until now. A couple basic how-to's I tried to read before somehow made it sound more complicated as well but your instructions are easy as hell to follow. 👌🏻 Any documentation you write on any project you'd make in the future will certainly be brilliant! Haha, thank you so much again!

**Reply by Serene-Arc** [reply]
Thanks for the kind words :) Using git as a developer can get more complex fast but just checking out code to build and use is usually a simple affair :) I'm glad I could help! If you're interested in learning git, I recommend the book *Version Control with Git* by Loeliger. That's the one that taught me pretty much everything about it.

**Reply by StuffonGithub** [reply]
@Serene-Arc Thanks! At some point, I'm gonna have to want to get back into trying to learn at least the basics of automation for the simple tasks so thanks for the reference!

I have an additional BDFR question if you don't mind, I didn't want to make a new topic as well. I can only run one instance of BDFR at a time, right? I'm getting a permission/in-use error if I run another but just wanted to confirm. 😁

**Reply by Serene-Arc** [reply]
At the moment, yes, because all instances of the BDFR look in the same place for the log file, they all try and write to it at once, which Windows doesn't allow at all. We're trying to figure out a solution to that as part of our testing, actually, since GitHub apparently automatically runs our integration tests in parallel. I'm thinking that one of the solutions for that problem would be to have another command-line argument that specifies the logfile location. If you want to open up an issue with the errors, I'll code up a quick fix soon.

**Reply by StuffonGithub** [reply]
Ah I see. I just wanted to confirm that the logfile permission error was intended and not just a wrong setting on my BDFR. I have no immediate need for multiple instances for now and I can make do with the current limitation. Don't wanna press on this issue and potentially backburner a lot of the more important stuff you have on your list.

But if you prefer to have an issue open for this feature for extra logging and referencing, I'd be more than happy to oblige. 😄

**Reply by Serene-Arc** [reply]
If you could open an issue, that would be swell, just to remind me and have something show up on my dashboard. It definitely wasn't *intended* per se, it just wasn't something that caused an issue. I developed the entirety of the v2 project on Linux, where the attitude is "if you want to do stupid shit, you can be stupid, I won't stop you". On Linux you can have multiple processes writing to the same file. It's a very bad idea and you'd get a scrambled logfile at best and gibberish as worst, but you can do it. Windows is the opposite, where it disallows it outright, so the error didn't appear until we used GitHub's CI to run it in a Windows environment. It's definitely something to fix.

**Reply by StuffonGithub** [reply]
Sorry for the delay. I created a Feature Request in Issues (#306). 😁

> "if you want to do stupid shit, you can be stupid, I won't stop you"

Haha I agree, I did play around Linux distros for a time and it was fun! Not much use for a basic user as myself though, so I tend to go with my main OSes as Windows and MacOS. But yeah, Windows is the type to want to simplify and have everything in a controlled environment given that majority of its users will be tech illiterate such as myself or just generally don't want the hassle. Nice that Github provides such tests outright though. Seems pretty helpful in cases of oversights. At the very least, you were aware of the issue before releasing.

**Reply by Serene-Arc** [reply]
Yeah, windows would never let you delete you're entire operating system with a single command :) Linux does though, and it's a reason that we've run into a lot of problems with the development. For example, Linux has exactly TWO characters that are not allowed in a file path: \0 and /. Windows...well, it would be easier to say what Windows *does* allow, which is pretty much ASCII, and even then a big chunk isn't allowed (things like ":/\<>" and so on. So saving filenames like that has been something we've had to work through, since it's different from each OS.

**Reply by StuffonGithub** [reply]
Yeah, I've noticed this subtle annoyance before when I had a Linux laptop. And I'm just realising now that it's weird they just don't agree on a universal system. At least, even just for simple file and folder names. Of course, there's a lot of nuances that I don't know about that obviously prevents implementation on any or all of them, otherwise they would have done so.

**Reply by Serene-Arc** [reply]
It's pretty simple, Windows is mostly hampered by the fact that they try to stay backwards compatible. When you're trying to stay compatible with code and practices from the 80s and 90s, you can get some weird stuff. Then there's, once again, Windows assuming no one knows how to use it. Characters like ></\?* are used in the command line to do things. you can use them in files on Linux, because the OS accommodates it and just escapes them. Not Windows though, for whatever reason, even though 95% of Windows users will ever open the command like.

One of the most annoying things is that in Windows files are case insensitive. So it registers TEST.png and test.png as the same thing, which can be a problem if you're trying to save things that only differ in that way. You get weird errors like the same thing trying to download on every run of the downloader.

**Reply by StuffonGithub** [reply]
Ah yeah, I remember reading some posts made by MS employees about how spaghetti the codes are for a lot of the main stuff in Windows like Explorer, Control Panel, and Task Manager. I think that's why even on the latest Win 10, we can still access the old Control Panel as well as the new 'Settings'.

Guess they prefer Windows being more consumer-friendly than devs. Tsk tsk.

**Comment by StuffonGithub** [comment]
Okay, I figured it out now. It's a simple Execution Policy error. But since entering the code manually in PS works as it should and running the .ps1 script just automatically closes the window after a second, no error message shows up and I had no way of knowing what was causing it. 🤦🏻‍♂️🤦🏻‍♂️🤦🏻‍♂️

Nevertheless, at least it's working how I prefer it to now and thanks so much for accommodating and helping me out, @aliparlakci and @Serene-Arc! ❤

**Comment by MuhammadEhsan2** [comment]


**Reply by Serene-Arc** [reply]
Hi. Please open a bug issue if there is no audio. If you have a question, please don't hijack another message thread, you can open one of your own.
